src/util/submit-addon.js

```diff
@@ -0,0 +1,388 @@
+        try {
+          const detailResponse = await this.fetch(detailUrl);
+          if (!detailResponse.ok) {
+            return reject(new Error('Getting addon details failed.'));
+          }
+          const detailResponseData = await detailResponse.json();
+
+          const file = detailResponseData.file;
+          if (file && file.status === 'public') {
+            clearTimeout(abortTimeout);
+            resolve(file.url);
+          } else {
+            // The add-on hasn't been approved yet, so wait for a while and try again.
+            approvalCheckTimeout = setTimeout(
+              pollApprovalStatus,
+              this.approvalCheckInterval
+            );
+          }
+        } catch (err) {
+          clearTimeout(abortTimeout);
+          reject(err);
+        }
+      };
+
+      pollApprovalStatus();
+    });
+  }
+
+  async getJson(response: typeof Response): Promise<any> {
+    if (response.status < 100 || response.status >= 500) {
+      return new Promise((resolve, reject) => {
+        reject(new Error(`Getting response failed: ${response.status}.`));
+      });
+    } else {
+      const data = await response.json();
+      return new Promise((resolve, reject) => {
+        if (!response.ok) {
+          this.logger.log(data);
+          reject(new Error('Bad Request.'));
+        } else {
+          resolve(data);
+        }
+      });
+    }
+  }
+
+  fetch(
+    url: string,
+    method: string = 'GET',
+    body?: typeof FormData | string,
+    jwt: typeof defaultJwt = defaultJwt
+  ): Promise<typeof Response> {
+    const authToken = jwt.sign({ iss: this.apiKey }, this.apiSecret, {
+      algorithm: 'HS256',
+      expiresIn: this.apiJwtExpiresIn,
+    });
+
+    this.logger.log(`Fetching URL: ${url}`);
+    let headers;
+    if (typeof body === 'string') {
+      headers = {
+        'Authorization': `JWT ${authToken}`,
+        'Accept': 'application/json',
+        'Content-Type': 'application/json',
+      };
+    } else {
+      headers = {
+        Authorization: `JWT ${authToken}`,
+        Accept: 'application/json',
+      };
+    }
+    return fetch(url, { method, body, headers });
+  }
+
+  async saveToFile(
+    contents: typeof Response.body, destPath: string): Promise<any> {
+    return promisify(pipeline)(contents, createWriteStream(destPath));
+  }
+
+  async downloadSignedFile(
+    fileUrl: string,
+    addonId: string
+  ): Promise<SignResult> {
+    const filename = fileUrl.split('/').pop(); // get the name from fileUrl
+    const dest = `${this.downloadDir}/${filename}`;
+    const response = await this.fetch(fileUrl);
+    if (!response.ok || !response.body) {
+      this.logger.log(`Download of signed xpi failed: ${response.status}.`);
+      throw new Error(`Downloading ${filename} failed`);
+    }
+    await this.saveToFile(response.body, dest);
+    return Promise.resolve({
+      success: true,
+      id: addonId,
+      downloadedFiles: [filename],
+    });
```

Comment:
No need to explicitly call Promise.resolve here, this is an async function and so its return value is always going to be a promise and the values we return from inside the function will be resolved by the returned promise, and so this can be just `return { ... };`.